bcm2711/sdio: Support for EMMC interfaces on the BCM2711

linguini1 opened the pull request:
## Summary

This initial implementation sets up support for the EMMC interfaces on the BCM2711. Only EMMC2 is tested since it is the interface of interest (connects to uSD card). MMCSD communication is functional and the boot partition of the SD card can be mounted and interacted with. Stress testing fails, insertion/removal events only work right after boot.

Closes #16948.
Closes #16949.

**NOTE:** Card insertion/removal events may not be possible with this implementation. I have followed the SD host controller simplified specification section 3.1 (SD card detection), but I have not gotten any results. I am leaving it alone for now.

**NOTE:** This is a draft PR for initial review. I still have some changes to make to stabilize this feature. I can't seem to get card insertion/removal interrupts to fire, although I see that they exist in the RPi Linux code and other projects. I've configured them the same way as other interrupts but no luck beyond initial boot.

**Does anyone know how to support 8-bit transfer?** The controller is capable of that, but the widebus functionality of the NuttX upper-half driver only seems to allow 1b or 4b modes.

## Impact

Users can now interact with the micro-SD card through NuttX! Right now they are able to mount the boot partition, however I will also add support for mounting two separate partitions on the micro SD card so that a disk filesystem exists.

## Testing

Tested using all the code included in this PR:

- Initialize SDIO lower-half properly
- Bind SDIO lower-half to MMCSD driver, verify that `/dev/mmcsd0` is mounted and that logs from the driver indicate correct operation
- Mount `/dev/mmcsd0` to NuttX at `/sd` and verify that I can see all of the files in the card with `ls`
- Verify that I can `cat` a file in the SD card
- Verify that I can `echo` to a file on the SD card to create a new one

### Successful mount

```
[    1.538000] fat_mount: FAT32:
[    1.541000] fat_mount:       HW  sector size:     512
[    1.546000] fat_mount:           sectors:         62521344
[    1.551000] fat_mount:       FAT reserved:        32
[    1.555000] fat_mount:           sectors:         62519296
[    1.561000] fat_mount:           start sector:    2080
[    1.565000] fat_mount:           root sector:     2
[    1.570000] fat_mount:           root entries:    0
[    1.575000] fat_mount:           data sector:     32608
[    1.580000] fat_mount:           FSINFO sector:   2049
[    1.584000] fat_mount:           Num FATs:        2
[    1.589000] fat_mount:           FAT sectors:     15264
[    1.594000] fat_mount:           sectors/cluster: 32
[    1.598000] fat_mount:           max clusters:    1952773
[    1.604000] fat_mount:       FSI free count       1952130
[    1.609000] fat_mount:           next free        1417

NuttShell (NSH) NuttX-12.11.0
nsh> ls 
/:
 dev/
 proc/
 sd/
nsh> ls /dev
/dev:
 console
 mmcsd0
 null
 zero
```

Listing the SD card boot partition:

```
/sd:
 bcm2711-rpi-4-b.dtb
 config.txt
 fixup4cd.dat
 fixup4.dat
 fixup4db.dat
 fixup4x.dat
 nuttx.bin
 start4cd.elf
 start4db.elf
 start4.elf
 start4x.elf
 overlays/
 test.txt
nsh> 
```

Using `cat` to see file contents of test.txt:

```
nsh> cat /sd/test.txt
This is a test! 
```

Using `echo` to append to test.txt:

```
nsh> echo "Another test!" >> /sd/test.txt
nsh> cat /sd/test.txt
This is a test!
Another test!
nsh> 
```

The `sdstress` test passes, so long as I configure it with **the byte parameter as 1023 or less. The moment I make it 1024 bytes or larger, I get a lot of errors.** I would imagine that this is something to do with the block gaps/the maximum supported transfer size being 1KB.

1023 byte stress test output:

```
nsh> sdstress -b 1023
Start stress test with 64 files, 1023 bytes and 32 iterations.
iteration 0 took 2607.000 ms: OK
iteration 1 took 1830.000 ms: OK
iteration 2 took 1566.000 ms: OK
iteration 3 took 3877.000 ms: OK
iteration 4 took 1608.000 ms: OK
iteration 5 took 1675.000 ms: OK
iteration 6 took 1605.000 ms: OK
iteration 7 took 1604.000 ms: OK
iteration 8 took 1602.000 ms: OK
iteration 9 took 1594.000 ms: OK
iteration 10 took 1704.000 ms: OK
iteration 11 took 1710.000 ms: OK
iteration 12 took 1664.000 ms: OK
iteration 13 took 1634.000 ms: OK
iteration 14 took 1650.000 ms: OK
iteration 15 took 1735.000 ms: OK
iteration 16 took 1658.000 ms: OK
iteration 17 took 1741.000 ms: OK
iteration 18 took 1661.000 ms: OK
iteration 19 took 1771.000 ms: OK
iteration 20 took 1690.000 ms: OK
iteration 21 took 1671.000 ms: OK
iteration 22 took 3241.000 ms: OK
iteration 23 took 2973.000 ms: OK
iteration 24 took 1914.000 ms: OK
iteration 25 took 2064.000 ms: OK
iteration 26 took 2490.000 ms: OK
iteration 27 took 1617.000 ms: OK
iteration 28 took 1616.000 ms: OK
iteration 29 took 1701.000 ms: OK
iteration 30 took 1633.000 ms: OK
iteration 31 took 1625.000 ms: OK
Test OK: Average time: 1897.844 ms
nsh> 
```

Output of stress test when above 1023 bytes:

```
nsh> sdstress -b 1024
Start stress test with 64 files, 1024 bytes and 32 iterations.
[  396.530000] mmcsd_readmultiple: ERROR: Card not ready: -110
[  396.534000] mmcsd_sendcmdpoll: ERROR: Wait for response to cmd: 0000004d failed: -110
read /sd/stress/[  396.593000] mmcsd_sendcmdpoll: ERROR: Wait for response to cmd: 0000004d failed: -110
tmp000 failed, r[  396.652000] mmcsd_sendcmdpoll: ERROR: Wait for response to cmd: 0000004d failed: -110
et: -1, errno 11[  396.711000] mmcsd_sendcmdpoll: ERROR: Wait for response to cmd: 0000004d failed: -110
0 -> Unknown err[  396.770000] mmcsd_sendcmdpoll: ERROR: Wait for response to cmd: 0000004d failed: -110
or 110
iteratio[  396.829000] mmcsd_sendcmdpoll: ERROR: Wait for response to cmd: 0000004d failed: -110
n 0 took 1067.00[  396.888000] mmcsd_sendcmdpoll: ERROR: Wait for response to cmd: 0000004d failed: -110
0 ms: FAIL
Test[  396.947000] mmcsd_sendcmdpoll: ERROR: Wait for response to cmd: 0000004d failed: -110
 FAIL: Average t[  397.006000] mmcsd_sendcmdpoll: ERROR: Wait for response to cmd: 0000004d failed: -110
ime: 33.344 ms
[  397.065000] mmcsd_sendcmdpoll: ERROR: Wait for response to cmd: 0000004d failed: -110
nsh> [  397.124000] mmcsd_sendcmdpoll: ERROR: Wait for response to cmd: 0000004d failed: -110
[  397.183000] mmcsd_sendcmdpoll: ERROR: Wait for response to cmd: 0000004d failed: -110
[  397.242000] mmcsd_sendcmdpoll: ERROR: Wait for response to cmd: 0000004d failed: -110
[  397.301000] mmcsd_sendcmdpoll: ERROR: Wait for response to cmd: 0000004d failed: -110
[  397.360000] mmcsd_sendcmdpoll: ERROR: Wait for response to cmd: 0000004d failed: -110
[  397.419000] mmcsd_sendcmdpoll: ERROR: Wait for response to cmd: 0000004d failed: -110
[  397.478000] mmcsd_sendcmdpoll: ERROR: Wait for response to cmd: 0000004d failed: -110
[  397.537000] mmcsd_writesingle: ERROR: Card not ready: -110
[  397.537000] file_put: ERROR: fs putfilep file_close() failed: -110
```

After this, it hangs. Following a reboot, here are the contents of the created `stress/` directory:

```
nsh> ls -l /sd/stress
/sd/stress:
 -rw-rw-rw-           0 tmp000
nsh> 
```

For whatever reason, very strange behaviour occurs when using a 64GB Sandisk card. Sometimes there is a data CRC failure after boot and the FAT filesystem never mounts, even when re-attempted, due to this error. Other times, the FAT filesystem mounts but attempting to create a file on the card causes `sendcmd` to be called with CMD13 repeatedly until the watchdog times out. I cannot figure out why this is happening for only the 64GB card.

linguini1 committed:
bcm2711/sdio: Support for EMMC interfaces on the BCM2711

This initial implementation sets up support for the EMMC interfaces on
the BCM2711. Only EMMC2 is tested since it is the interface of interest
(connects to uSD card). MMCSD communication is functional and the boot
partition of the SD card can be mounted and interacted with.
Insertion/removal interrupts do not fire after initial boot, and
sdstress fails with any byte size larger than 1023. 32GB card works
perfectly, while 64GB card exhibits strange behaviour when interacting
with VFAT filesystem.

Signed-off-by: Matteo Golin <matteo.golin@gmail.com>

acassis reviewed (commented):
Nice work @linguini1 ! Kudos!!!

acassis reviewed (commented):
Please update the Documentation/ to include this new feature. Also include board profile "sd" example explaining how to use

linguini1 committed:
docs/platforms/arm64/bcm2711: microSD documented, improvements

The documentation is now updated to reflect the new EMMC support. I have
also improved the format to match the standard board documentation
template, and included better support descriptions for features that
have varying levels of implementation.

Signed-off-by: Matteo Golin <matteo.golin@gmail.com>

acassis reviewed (dismissed):
I'm approving since it works for some SDCards, but maybe you should include "depends on EXPERIMENT" to BCM2711_EMMC. This way users will reduce the expectations on this driver.

cederom reviewed (approved):
Thank you @linguini1 amazing work, great description, documentation, and testing!! :-)

I have ordered the boards and will be able to test next week with rPI 4B 8GB and SanDisk HighEndurance 64GB and Patriot EP 64GB cards but you have my go already :-)

linguini1 commented:
> Thank You for porting NuttX to rPI 4B @linguini1 this is hard and amazing work kudos!! :-)
> 
> I have now setup and I know how to use it let me know when you need more testing :-)
> 
> During this tests I found some related improvements:
> * `free -h` to get human readable output, created issue https://github.com/apache/nuttx/issues/17274.
> * `sdstress` does not work correctly when `/sd/stress` exists, created issue https://github.com/apache/nuttx-apps/issues/3205.

I really appreciate your time testing! Great finds as well :)
